lighthouse-core/audits/seo/font-size.js

```diff
@@ -0,0 +1,234 @@
+  if (styleDeclaration && styleDeclaration.type === CSSStyleDeclaration.Type.Regular) {
+    const startLine = styleDeclaration.range ? styleDeclaration.range.startLine : 0;
+    const startColumn = styleDeclaration.range ? styleDeclaration.range.startColumn : 0;
+    return `${styleDeclaration.styleSheetId}@${startLine}:${startColumn}`;
+  } else {
+    return `node_${node.nodeId}`;
+  }
+}
+
+class FontSize extends Audit {
+  /**
+   * @return {!AuditMeta}
+   */
+  static get meta() {
+    return {
+      name: 'font-size',
+      description: 'Document uses legible font sizes.',
+      failureDescription: 'Document doesn\'t use legible font sizes.',
+      helpText: 'Font sizes less than 16px are too small to be legible and require mobile ' +
+      'visitors to “pinch to zoom” in order to read. ' +
+      '[Learn more](https://developers.google.com/speed/docs/insights/UseLegibleFontSizes).',
+      requiredArtifacts: ['FontSize', 'URL', 'Viewport'],
+    };
+  }
+
+  /**
+   * @param {!Artifacts} artifacts
+   * @return {!AuditResult}
+   */
+  static audit(artifacts) {
+    const hasViewportSet = ViewportAudit.audit(artifacts).rawValue;
+    if (!hasViewportSet) {
+      return {
+        rawValue: false,
+        debugString: 'Text is illegible because of a missing viewport config',
+      };
+    }
+
+    const {
+      failingNodesData,
+      totalTextLength,
+      failingTextLength,
+      visitedTextLength,
+    } = artifacts.FontSize;
+
+    if (totalTextLength === 0) {
+      return {
+        rawValue: true,
+      };
+    }
+
+    const failingRules = getUniqueFailingRules(failingNodesData);
+    const percentageOfPassingText =
+      (visitedTextLength - failingTextLength) / visitedTextLength * 100;
+    const pageUrl = artifacts.URL.finalUrl;
+
+    const headings = [
+      {key: 'source', itemType: 'url', text: 'Source'},
+      {key: 'selector', itemType: 'code', text: 'Selector'},
+      {key: 'coverage', itemType: 'text', text: 'Coverage'},
+      {key: 'fontSize', itemType: 'text', text: 'Font Size'},
+    ];
+
+    const tableData = failingRules.sort((a, b) => b.textLength - a.textLength)
+      .map(({cssRule, textLength, fontSize, node}) => {
+        const percentageOfAffectedText = textLength / visitedTextLength * 100;
+        const origin = findStyleRuleSource(pageUrl, cssRule, node);
+
+        return {
+          source: origin.source,
+          selector: origin.selector,
+          coverage: `${percentageOfAffectedText.toFixed(2)}%`,
+          fontSize: `${fontSize}px`,
+        };
+      });
+
+    // all failing nodes which were not fully analyzed will be shown in the 'Other' category
+    const analyzedFailingTextLength =
+      failingRules.reduce((sum, {textLength}) => sum += textLength, 0);
+    if (analyzedFailingTextLength < failingTextLength) {
+      const percentageOfUnanalyzedFailingText =
+        (failingTextLength - analyzedFailingTextLength) / visitedTextLength * 100;
```

Comment:
Not sure I totally follow this calculation, doesn't `tableData` already contain the total sum of `visitedTextLength`?

this is pretty complicated by the double opportunity for analysis to fall off with nodes, so maybe I'll just gut check what each is trying to capture real quick

for each node that we visited *and* got the css rule for, we'll have a distinct row in the table data
for all other nodes that we visited but did not get css rule for, we'll have an "Other" row in the table data
to account for all the text that we didn't visit at all, we'll show a debugString highlighting that this was only a sample

does that sound right to you? if so, I'm a little fuzzy on how the non-css rule ones are getting filtered out from `failingRules`. is that by virtue of the `getUniqueFailingRules`?

Reply:
It definitely got a bit complicated, but your gut check is all correct 👌

>  how the non-css rule ones are getting filtered out from `failingRules`. is that by virtue of the `getUniqueFailingRules`?

Only analyzed nodes make it to the `failingRules` because gatherer returns only analyzed nodes via `failingNodesData`. Everything else is calculated based on text length stats (`totalTextLength`, `failingTextLength`, `visitedTextLength`). Perhaps `failingNodesData` should be called `analyzedFailingNodesData`?

At first, I did try to pass all failing visited nodes via `failingNodesData` and calculate "Other" row in audit, but eventually figured out that passing only analyzed nodes + length stats makes everything much simpler.

Reply:
ah ok, I think I got it. Two proposals then

* let's rename failingNodesData as you suggest since the similarity to `failingTextLength` threw me off a bit :)
* let's compute the `analyzedFailingTextLength` in the gatherer too